samples/emoji-search/presenter/src/commonTest/kotlin/com/example/redwood/emojisearch/presenter/ManualCoroutineDispatcher.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2023 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.redwood.emojisearch.presenter
+
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Runnable
+
+/** Extremely basic dispatcher that requires [executeQueuedJobs] to be invoked manually. */
+internal class ManualCoroutineDispatcher : CoroutineDispatcher() {
```

Comment:
The `CoroutineDispatcher` I’m using here is super naive and likely to cause problems. Is there something built-in that works similarly?

Reply:
Yeah you don't want to be in the dispatcher business. This one has non-determinism with anyone that uses libraries which run on other dispatchers and it infinite loops on self-updating effects. I would delete this and use a normal one.

My intuition is that upon a snapshot, the best thing to do is tick the clock until the applier signals that it has new changes. And then when you get that snapshot, to ensure it actually differs from the previous one (and if not keep sending frames). Do this until a timeout after, say, 2 seconds, or 10.

This ensures that each call to snapshot produces a new, distinct snapshot which reflects some change. Beyond that, though, you still likely will have some non-determinism unless you rigorously control all of the dispatchers used by libraries.

I think a better user API is to completely ignore the concept of frames, and instead focus on the desired conditions and performing actions. This is in line with how a user would behave on the real display. 

For example:
- Wait for "email" input field to appear
- Enter email
- Enter password
- Click submit button
- Wait for payment pad to appear

Behind the scenes this will feature the same automatic frame clock ticking and waiting for the applier to signal changes have occurred, but then there's an additional step of ensuring that it's a snapshot which meets the requirements of the caller. And if we can't find one within a reasonable timeout the test fails with the latest snapshot and the missing condition.

This is the pattern we are encouraging with regular UI tests and it's why we are deprecating RxIdler and okhttp-idling-resource. You simply can't get reliable determinism by trying to monitor running work. And while you're testing Compose logic and value types, fundamentally is _is_ the same as a UI test since we normally project those values directly into platform-specific UI.

Reply:
One nice advantage we have here is that there’s no concurrency unless we introduce it.

But I take your point that we don’t want to be in the dispatcher business, not at all. I think I’ll just punt that to the user, and make them inject whatever dispatcher they want.

Reply:
Alright, I’ve added code that lets callers await a distinct snapshot.

I decided to introduce a `ChangeTracker` thing to signal when changes were applied.

The only thing I’m worried about is tearing, where a snapshot is observed mid-frame.  I don‘t think this can occur in practice, since that coroutine would need to suspend mid-frame. Instead I suspect the `Semaphore` I’m using won’t resume until the composition is idle.

Reply:
> I decided to introduce a `ChangeTracker` thing to signal when changes were applied.
>
> The only thing I’m worried about is tearing, where a snapshot is observed mid-frame. I don‘t think this can occur in practice, since that coroutine would need to suspend mid-frame. Instead I suspect the `Semaphore` I’m using won’t resume until the composition is idle.

The `WidgetApplier` already has a callback to tell you when changes have been applied.

Here's how the protocol-based widgets work:
```kotlin
val applier = WidgetApplier(bridge.provider, bridge.root) {
  bridge.createDiffOrNull()?.let(diffSink::sendDiff)
}
val composition = RedwoodComposition(scope, applier)
```

> One nice advantage we have here is that there’s no concurrency unless we introduce it.

We tried this with Molecule and it failed miserably. We can try it again if you want, but I think once you move away from testing samples to testing real-world code it becomes much more challenging to accomplish this.

Reply:
Ooooh I was looking for the change signalling API and couldn’t find it. This is much better, especially 'cause I get one batch of changes.